modules/tobira/src/main/java/org/opencastproject/tobira/impl/TobiraApi.java

```diff
@@ -103,79 +138,154 @@ public void setSearchService(SearchService service) {
+            // same reasoning applies the other way: if we have more series than `preferredAmount`,
+            // then all events with modified dates after the modified date of the last series we
+            // return will be included in the next request anyway. So we might as well filter them
+            // out here to save a bit on the size of this request.
+            //
+            // The reason we first filter series, then events, is because it is likely that an
+            // Opencast instance contains way more events than series. This means that
+            // `hasMoreSeriesInRange` being true is actually very uncommon.
+            if (!hasMoreSeriesInRange) {
+              return true;
             }
+
+            final Date lastSeriesModifiedDate = rawSeries.get(rawSeries.size() - 1)
+                .getModifiedDate();
+            return !event.getModified().after(lastSeriesModifiedDate);
           })
+          .map(event -> new Item(event));
+
+      final Stream<Item> seriesItems = rawSeries.stream()
+          .limit(preferredAmount)
+          .map(series -> new Item(series));
+
+
+      // Combine series and events into one combined list and sort it.
+      //
+      // The sorting is, again, not for correctness, because consumers of this API need to be
+      // able to deal with that. However, sorting this here will result in fewer temporary objects
+      // or invalid states in the consumer.
+      final ArrayList<Item> items = Stream.concat(eventItems, seriesItems)
           .collect(Collectors.toCollection(ArrayList::new));
+      items.sort(Comparator.comparing(item -> item.getModifiedDate()));
+
+
+      // Obtain information to allow Tobira to plan the next harvesting request.
+      boolean hasMore = hasMoreEvents || hasMoreSeriesInRange;
+      final Date includesItemsUntilRaw;
+      if (!hasMoreEvents && !hasMoreSeriesInRange) {
+        // All events and series up to now have been harvested.
+        includesItemsUntilRaw = new Date();
+      } else if (!hasMoreEvents && hasMoreSeriesInRange) {
+        // `rawEvents` contains all events that currently exist. Our response won't contain the ones
+        // that have a modified date after the one of the last raw series. But that means that we
+        // will return all series and events until the last raw series.
+        includesItemsUntilRaw = rawSeries.get(rawSeries.size() - 1).getModifiedDate();
+      } else if (hasMoreEvents && !hasMoreSeriesInRange) {
+        // There are more events, but no additional series in the range from `since` to the modified
+        // date of the last raw event. So we know there are no other events or series before the
+        // last raw events.
+        includesItemsUntilRaw = rawEvents[rawEvents.length - 1].getModified();
+      } else {
+        // There are more events and more series in the given range. In theory, this would be
```

Comment:
Not only related to this part of the code, but we spend a lot of effort on doing things to figure out if there are more events, … do we need that at all? Long-term, the harvesting should only be the fallback option, since we also want synchronous updates. Even if we don't know that there are more events, we will harvest again eventually. The limit is mostly for an initial harvest, so we do not load ten thousand events at once. But we know about the initial harvest and can have admins just increase the harvesting speed during that period if necessary.

So, this kind of feels a bit overengineered and kind of feels like code where someone may make the wrong turn eventually, causing an error.

Reply:
That's a very good point and that thought crossed my mind, too. If we keep the current logic, I certainly want to have tests for the various edge cases. And yes, the logic got a bit more involved than I anticipated. I was very close to just giving up on smart logic entirely, like you suggest.

However, right now I'm fairly happy with the code and the explaining comments. Not being smart about this at all does have a few disadvantages. It would mean transferring lots of duplicate series, sending stuff out of order very often (which Tobira needs to deal with anyway, BUT leads to temporary object and extra handling on Tobira side each time), and sending lots of items more than once. It's kind of nice if the harvesting can say "no new data" instead of always sending the last item again and again.

So yeah, I also have my troubles with the complexity this adds. Right now, I have a very small tendency to be smart here. Especially since we now already have the code and I don't expect to change it in the near future. But you'd certainly be able to convince me otherwise.

---

Final note about the synchronous API: There is still the idea in my head that Tobira merely offers one endpoint that, when requested, just queues a new harvest right away. Ala "hey, there is new data, please harvest now!". Sure, in the worst case, that would mean twice as many HTTP requests (though one would contain no data, i.e. is small). But this would be easy to add, would basically offer the same advantages as a "push data" API, and most importantly would avoid extra logic and complexities.

Reply:
> It's kind of nice if the harvesting can say "no new data" instead of always sending the last item again and again.

Remind me again: Why would we need to send one item?